ibis/expr/types.py

```diff
@@ -1477,99 +856,65 @@ def sequence(values):
-
-    def __getitem__(self, key):
-        return self.values[key]
-
-    def __add__(self, other):
-        other_values = getattr(other, 'values', other)
-        return type(self.op())(self.values + other_values).to_expr()
+    return ops.ValueList(values).to_expr()
 
-    def __radd__(self, other):
-        other_values = getattr(other, 'values', other)
-        return type(self.op())(other_values + self.values).to_expr()
 
-    def __bool__(self):
-        return bool(self.values)
+def as_value_expr(val):
+    import pandas as pd
+    if not isinstance(val, Expr):
+        if isinstance(val, (tuple, list)):
+            val = sequence(val)
+        elif isinstance(val, pd.Series):
+            val = sequence(list(val))
+        else:
+            val = literal(val)
 
-    def __len__(self):
-        return len(self.values)
+    return val
 
-    def type(self):
-        from ibis.expr import rules
-        return rules.highest_precedence_type(self.values)
 
+_parameter_counter = itertools.count()
 
-class SortExpr(Expr):
 
-    def _type_display(self):
-        return 'array-sort'
+def param(type, name=None):
+    """Create a parameter of a particular type to be defined just before
+    execution.
 
+    Parameters
+    ----------
+    type : dt.DataType
+        The type of the unbound parameter, e.g., double, int64, date, etc.
+    name : str, optional
+        The name of the parameter
 
-class ValueList(ValueOp):
+    Returns
+    -------
+    ScalarExpr
 
+    Examples
+    --------
+    >>> import ibis
+    >>> import ibis.expr.datatypes as dt
+    >>> start = ibis.param(dt.date)
+    >>> end = ibis.param(dt.date)
+    >>> schema = [('timestamp_col', 'timestamp'), ('value', 'double')]
+    >>> t = ibis.table(schema)
+    >>> predicates = [t.timestamp_col >= start, t.timestamp_col <= end]
+    >>> expr = t.filter(predicates).value.sum()
     """
-    Data structure for a list of value expressions
-    """
-
-    def __init__(self, args):
-        self.values = list(map(as_value_expr, args))
-        super(ValueList, self).__init__(self.values)
+    import ibis.expr.datatypes as dt
+    import ibis.expr.operations as ops
 
-    def root_tables(self):
-        return distinct_roots(*self.values)
-
-    def _make_expr(self):
-        return ListExpr(self)
+    if name is None:
+        name = 'param_{:d}'.format(next(_parameter_counter))
 
+    expr = ops.ScalarParameter(dt.dtype(type), name=name).to_expr()
+    return expr.name(name)
 
-def bind_expr(table, expr):
-    if isinstance(expr, (list, tuple)):
-        return [bind_expr(table, x) for x in expr]
 
-    return table._ensure_expr(expr)
+class UnnamedMarker(object):
+    pass
 
 
-def find_base_table(expr):
-    if isinstance(expr, TableExpr):
-        return expr
-
-    for arg in expr.op().flat_args():
-        if isinstance(arg, Expr):
-            r = find_base_table(arg)
-            if isinstance(r, TableExpr):
-                return r
+unnamed = UnnamedMarker()
```

Comment:
I remember deleting this. Am I mistaken? If not, why bring it back?

Reply:
Probably, it's present in the [master](https://github.com/ibis-project/ibis/blob/a2a9f0ee6d743859b043be88ed88c791950200e1/ibis/expr/types.py#L1348)